components/Rating.js

```diff
@@ -103,20 +103,50 @@ class Rating extends Component {
     });
   }
 
+  /* Calculate height of all labels and store in this.state.labelHeight[] in one call. 
+     Ratings that start hidden will have no labelRefs.
+     The labelRefs are only set during their render by the _label() function when made visible.
+     This function is optimised for use in componentDidMount. */
+  setLabelHeightsInitially() {
+    var tmp_labelHeights = [];
+    // the labelRefs are only set during their render by the _label() function
+    var numberOfLabels=this.labelRefs.length;
+    
+    // for some rating types this property will be === 'undefined' which is ok as they have no labels anyway :)
+    if(this.props.data.properties.showCaptions) {
+      if(this.state.labelHeight.length < this.labelRefs.length) {
+        this.labelRefs.forEach( (ref,i) => {
+          if(i>=tmp_labelHeights.length) {
+            this.refs[ref]._component.measure((ox, oy, width, height, px, py) => {
+              // if a label is not visible, the height will be 0. Don't set labelHeights for invisible items!
+              if(height>0) { 
+                tmp_labelHeights[i]=height+1; // add margin=1 above text
+              }
+              // only change state when heights of all labels are known.
+              if(tmp_labelHeights.length == numberOfLabels) {
+                this.setState((prevState) => {
+                  return {
+                    labelHeight:tmp_labelHeights
+                  }
+                })
+              }
+            })
+          }
+        })
+      }
+    }
+  }
+
   componentDidMount() {
     setTimeout(() =>  {
-      this.labelRefs.forEach( (ref,i) => {
-        this.refs[ref]._component.measure((ox, oy, width, height, px, py) => {
-          this.setState((prevState) => {
-            return {
-              labelHeight:Object.assign([...prevState.labelHeight],{[i]:height})
-            }
-          })
-        })
-      })
+      this.setLabelHeightsInitially();
     });
   }
 
+  componentDidUpdate() {
+    this.setLabelHeightsInitially();
```

Comment:
What's the use of calling the function in `componentDidUpdate` if the function never does anything anymore due to line 117?

Reply:
That is for the labels of initially hidden Rating blocks. The componentDidUpdate() will catch those when they become visible.